item-requests.md

```diff
@@ -16,26 +16,84 @@ Example: to be provided
 Provide detailed information about a specific request. The JSON response document will resemble this:
 ```json
 {
-  "id": "12ef",
-  "token": "56cd",
-  "allfiles": true,
-  "request_email": "jqpublic@example.com",
-  "request_name": "John Q. Public",
-  "request_date": "20180718T205000",
-  "accept_request": false,
-  "decision_date": "20180718T295100",
-  "expires": "20181231T235959",
-  "request_message": "Please send this to me.",
-  "item_id": "36ab",
-  "bitstream_id": "44cd"
+  "id":1,
+  "decisionDate":null,
+  "expires":null,
+  "requestDate":"2021-09-17T14:55:50.089+00:00",
+  "token":"c19e820ea0b270f0ec98323864dcc8b8",
+  "acceptRequest":false,
+  "allfiles":false,
+  "type":"itemrequest",
+  "bitstreamId":"dca2dadb-7028-40a9-ab39-1a47726df7ef",
+  "itemId":"71d9fb0c-cc36-41c1-b1d3-63887b414fca",
+  "requestEmail":"jsmith@example.com",
+  "requestMessage":"Please send me a copy of this.",
+  "requestName":"John Smith",
+  "_links":{
+    "bitstream":{
+      "href":"http://localhost/api/tools/itemrequests/1/bitstream"
+    },
+    "item":{
+      "href":"http://localhost/api/tools/itemrequests/1/item"
+    },
+    "self":{
+      "href":"http://localhost/api/tools/itemrequests/1"
+    }
+  }
 }
 ```
 
+Item properties:
+
+  * type: always "itemrequest".  READ-ONLY
+  * id: internal unique identifier of the request.  READ-ONLY
+  * itemId: UUID of the requested Item.
+  * bitstreamId: UUID of the requested bitstream.
+  * allfiles: true if the request is for all bitstreams of the item.
+  * requestEmail: email address of the person requesting the files.
+  * requestName: Human-readable name of the person requesting the files.
+  * requestMessage: arbitrary message provided by the person requesting the files.
+  * requestDate: date that the request was recorded.  READ-ONLY.
+  * expires: date on which the request is considered expired.
+  * acceptRequest: true if the request has been granted.
+  * decisionDate: date that the request was granted or denied.  READ-ONLY.
+  * token: opaque string which uniquely identifies this request.  READ-ONLY
+
 Exposed links:
 
   * item: the item requested
   * bitstream: the bitstream requested
 
+## Creating a Request
+**POST /api/tools/itemrequests**
+
+Anyone may create an item request.  The Content-Type is JSON.  Example:
+```json
+{
+    "itemId": "71d9fb0c-cc36-41c1-b1d3-63887b414fca",
+    "allfiles": false,
+    "bitstreamId": "dca2dadb-7028-40a9-ab39-1a47726df7ef",
+    "requestEmail": "jqpublic@example.com",
+    "requestName": "John Q. Public",
+    "requestMessage": "Please send this to me."
+}
+```
+The response contains the complete request in JSON format, as shown in Single Request.
+
+An appropriate person will be notified that the request has been filed.
```

Comment:
Similar to other REST Contract docs, I'd recommend we include a bulleted list of "Return Codes" possible and what they each mean.  You can see an example in the Collection Creation docs here: https://github.com/DSpace/RestContract/blob/main/collections.md#creating-a-collection

For this endpoint, based on reviewing the code (in https://github.com/DSpace/DSpace/pull/2129), it looks like the possible return codes would be `200 OK` (success), `422 UNPROCESSABLE ENTITY` (if bad/missing data sent), and `404` if item not found.

Reply:
Is there a table somewhere showing which exceptions and other conditions map to which return codes?

Reply:
@mwoodiupui : We don't have a table showing the mapping, but it *is* pretty much controlled by a single class. This one: https://github.com/DSpace/DSpace/blob/main/dspace-server-webapp/src/main/java/org/dspace/app/rest/exception/DSpaceApiExceptionControllerAdvice.java

There are a few exceptions not listed in there as they are automated more at the Spring Security level (usually things like 403s or similar).  But, that class provides mappings of our own Java level exceptions to HTTP error codes.

Reply:
Return codes added.

Reply:
I've made a table for my own use, and would be happy to contribute it somewhere.

Reply:
@mwoodiupui : Thanks. I'll admit, I'm not sure offhand where the best place would be to create such a table.  Our DSpace 7 REST API *developer* docs are a bit lacking...there are some older pages here: https://wiki.lyrasis.org/display/DSPACE/DSpace+7+REST+API

I guess one option might be to simply document it in the codebase itself in this README? https://github.com/DSpace/DSpace/blob/main/dspace-server-webapp/README.md

Obviously though it should also link to `DSpaceApiExceptionControllerAdvice` class, in case updates have been made to that class which are not yet in the table.